File: typescript/packages/core/src/in-memory-session.ts
import type { Message } from './types/domain.js';
import type { SessionRecord, SessionSearchHit, SessionStore } from './types/contracts.js';

function messageText(message: Message): string {
  return typeof message.content === 'string' ? message.content : JSON.stringify(message.content);
}

export class InMemorySessionStore implements SessionStore {
  private sessions = new Map<string, SessionRecord>();

  async get(sessionId: string): Promise<SessionRecord | null> {
    return this.sessions.get(sessionId) ?? null;
  }

  async save(session: SessionRecord): Promise<void> {
    this.sessions.set(session.id, session);
  }

  async appendMessages(sessionId: string, messages: Message[]): Promise<void> {
    const existing = this.sessions.get(sessionId);
    const now = new Date().toISOString();
    if (!existing) {
      this.sessions.set(sessionId, {
        id: sessionId,
        messages: [...messages],
        createdAt: now,
        updatedAt: now,
      });
      return;
    }
    existing.messages.push(...messages);
    existing.updatedAt = now;
  }

  async list(): Promise<SessionRecord[]> {
    return [...this.sessions.values()].sort((a, b) => a.id.localeCompare(b.id));
  }

  async search(query: string, limit = 20): Promise<SessionSearchHit[]> {
    const normalized = query.toLowerCase();
    const hits: SessionSearchHit[] = [];

    for (const session of this.sessions.values()) {
      session.messages.forEach((message, index) => {
        const text = messageText(message);
        if (text.toLowerCase().includes(normalized)) {
          hits.push({
            sessionId: session.id,
            messageIndex: index,
            snippet: text.slice(0, 120),
          });
        }
      });
    }

    return hits.slice(0, limit);
  }
}

File: python/src/agent_framework/core/in_memory_session.py
from __future__ import annotations

import json
from dataclasses import asdict
from datetime import datetime, timezone

from agent_framework.core.contracts import SessionRecord, SessionStore
from agent_framework.core.domain import Message


def _message_text(message: Message) -> str:
    if isinstance(message.content, str):
        return message.content
    return json.dumps([asdict(part) for part in message.content])


class InMemorySessionStore(SessionStore):
    def __init__(self) -> None:
        self._sessions: dict[str, SessionRecord] = {}

    async def get(self, session_id: str) -> SessionRecord | None:
        return self._sessions.get(session_id)

    async def save(self, session: SessionRecord) -> None:
        self._sessions[session.id] = session

    async def append_messages(self, session_id: str, messages: list[Message]) -> None:
        now = datetime.now(timezone.utc).isoformat()
        existing = self._sessions.get(session_id)
        if existing is None:
            self._sessions[session_id] = SessionRecord(
                id=session_id,
                messages=list(messages),
                created_at=now,
                updated_at=now,
            )
            return
        existing.messages.extend(messages)
        existing.updated_at = now

    async def search(self, query: str, limit: int = 20) -> list[dict[str, str | int]]:
        normalized = query.lower()
        hits: list[dict[str, str | int]] = []
        for session in self._sessions.values():
            for index, message in enumerate(session.messages):
                text = _message_text(message)
                if normalized in text.lower():
                    hits.append(
                        {
                            "session_id": session.id,
                            "message_index": index,
                            "snippet": text[:120],
                        }
                    )
        return hits[:limit]
